Mathlib/Topology/MetricSpace/Ultra/TotallyDisconnected.lean

```diff
@@ -0,0 +1,33 @@
+/-
+Copyright (c) 2024 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Topology.MetricSpace.Defs
+import Mathlib.Topology.MetricSpace.Ultra.Basic
+
+/-!
+## Ultrametric spaces are totally disconnected
+
+In a metric space with an ultrametric, the space is totally disconnected.
+
+## Tags
+
+ultrametric, nonarchimedean, totally disconnected
+-/
+open Metric IsUltrametricDist
+
+instance {X : Type*} [MetricSpace X] [IsUltrametricDist X] : TotallyDisconnectedSpace X :=
+  totallyDisconnectedSpace_iff_connectedComponent_subsingleton.mpr fun x =>
+  suffices IsTotallyDisconnected (connectedComponent x) from
+    this _ Set.Subset.rfl isPreconnected_connectedComponent; by
+    intro t _ ht' y hy z hz
+    contrapose! ht'
+    intro H
+    specialize H (ball y (dist y z / 2)) (ball y (dist y z / 2))ᶜ
+      isOpen_ball (isClosed_ball _ _).isOpen_compl
+    simp only [Set.union_compl_self, Set.subset_univ, Set.inter_compl_self, Set.inter_empty,
+      Set.not_nonempty_empty, imp_false, true_implies] at H
+    refine H ⟨y, hy, mem_ball.mpr ?_⟩ ⟨z, hz, ?_⟩
+    · simpa using ht'
+    · simp [dist_comm, dist_nonneg]
```

Comment:
I find this proof a bit hard to read, not least because of starting a big tactic block inside the argument to a lemma  Can you rephrase it slightly and maybe add a comment or two to explain what's going on? For a start, you can get rid of the first few lines by starting with `rw [totallyDisconnectedSpace_iff]`, i.e.
```
instance {X : Type*} [MetricSpace X] [IsUltrametricDist X] : TotallyDisconnectedSpace X := by
  rw [totallyDisconnectedSpace_iff]
  intro t _ ht' y hy z hz
  contrapose! ht'
  intro H
  specialize H (ball y (dist y z / 2)) (ball y (dist y z / 2))ᶜ
    isOpen_ball (isClosed_ball _ _).isOpen_compl
  simp only [Set.union_compl_self, Set.subset_univ, Set.inter_compl_self, Set.inter_empty,
    Set.not_nonempty_empty, imp_false, true_implies] at H
  refine H ⟨y, hy, mem_ball.mpr ?_⟩ ⟨z, hz, ?_⟩
  · simpa using ht'
  · simp [dist_comm, dist_nonneg]
```
works, but further simplification is surely possible.

It feels to me like the "core" of this argument can be factored out, something like "a space is totally disconnected if for any distinct x, y there is a clopen containing x but not y". Maybe it would be useful to make this explicit.

Reply:
> It feels to me like the "core" of this argument can be factored out, something like "a space is totally disconnected if (iff?) for any distinct x, y there is a clopen containing x but not y". Maybe it would be useful to make this explicit.

Wait, that is in the library already. So one can golf this down to
```lean
instance {X : Type*} [MetricSpace X] [IsUltrametricDist X] : TotallyDisconnectedSpace X := by
  refine (totallyDisconnectedSpace_iff X).mpr (isTotallyDisconnected_of_isClopen_set fun x y h ↦ ?_)
  obtain ⟨r, hr, hr'⟩ := exists_between (dist_pos.mpr h)
  refine ⟨_, IsUltrametricDist.isClopen_ball x r, ?_, ?_⟩
  · simp only [mem_ball, dist_self, hr]
  · simp only [mem_ball, dist_comm, not_lt, hr'.le]
```
(Note the use of "exists_between" rather than explicitly halving the distance, this is a useful trick.)

Reply:
Will move to a separate PR.

Reply:
#17297